### DISCUSSION TITLE
Plugin Ideas

### DISCUSSION THREAD

**Discussion by aaronr8684** [txoof/PaperPi#57] [Ideas]
A central location for tracking new plugins and ideas

Using this should keep the issues list down

**Comment by aaronr8684** [comment]
A few ideas I've had:
- Full Screen Clock (probably just a layout on the existing clock plugin)
- Alarm Clock
- Todoist Integration (and others)
- Stock/Crypto with more layout options - e.g. https://github.com/donbing/bitbot
- Log Tracking Display (open and display last XX number of lines)
- File Display (similar to log, but for directories)
- METAR / TAF / NOTAM / SIGMET
- Daily quote / scripture
- GitHub Tracker
- Twitter / Reddit Feed
- Morning Dashboard (Weather, To-Do/Appts, Time) - e.g. https://github.com/mendhak/waveshare-epaper-display

**Comment by aaronr8684** [comment]
Ideas from the issues tab:
- Daily Quote - https://github.com/veebch/stonks
- Photo Slideshow (Imgur, r/blackandwhite, Google Image, etc.)
- Stock / Crypto with multiple per layout
- News Headlines (NYT, NPR, BBC, etc.)
- Lyrics - https://github.com/TimoLin/SpotifyLyrics
- VSMP - https://github.com/TomWhitwell/SlowMovie

**Comment by thoughtfix** [comment]
I would REALLY love that slideshow plugin. I could probably write some Python+Raspberry Pi web UI to allow users to upload images, but I have no idea if this plugin needs a specific resolution/color depth, how to write it to the display, and how to cycle it.

**Reply by txoof** [reply]
@thoughtfix 
This should be pretty easy to manage. PaperPi handles all the image processing for you. Right now only grayscale images or 1 bit dithered images are supported. The HD screens support grayscale, otherwise the images are converted to 1 bit dithered images. 

All you need to create is a little bit of code that scans a directory and uses whatever logic you choose to pick an image from the directory.  Take a look at the [demo_plugin](https://github.com/txoof/epd_display/tree/master/paperpi/plugins/demo_plugin) or the [moon_phase](https://github.com/txoof/epd_display/tree/master/paperpi/plugins/moon_phase) plugin to get an idea of how everything fits together. 

If you want to collaborate on this, shoot me a line. I can help with some of the more challenging elements like creating the layout file. 

The basic application flow is this:

- PaperPi cycles to the next active plugin
- The `update_function()` is triggered
- The `update_function()` returns a dictionary that includes text or images for each of the defined fields (specified in a layout file)
- PaperPi does all the necessary calculations to scale the text and images to the resolution and ratio of the screen
- PaperPi creates an image based on those calculations 
- PaperPi writes the image to the screen doing whatever refreshes are required for the particular screen
- Rinse, repeat previous steps indefinitely

**Reply by txoof** [reply]
I should say, full color images are supported, but only grayscale/1bit images are displayed. We don't support any color features right now.

**Comment by thoughtfix** [comment]
I am fortunate to have an HD greyscale 6 inch display, which works with this software (yay!) and I just finished another Python project that does random selection of a file for display over SPI, but it's a whole lot cuter https://www.youtube.com/shorts/4nAN1jXC2cE 

I'll dig in on these examples. I've never used Jupyter before, but if I can figure it out I'll make a PR and even CC license a couple cat pictures for free use in a demo. if I make a web server that is specific for storing images for this and my other project, I'll put it on GH and share the link with you.

THANK YOU for doing all the scaling! For my other project, I had to manually crop my images to squares and use a script to resize them.

**Reply by txoof** [reply]
Cool!

You don't need to work in Jupyter if you don't want to. The intention was to remove all the Jupyter from the source and keep it all pure python. 

Clearly some got left behind 😳.

You should be able to just make a copy of the demo plugin and work around that.l to get started.

Look forward to a pr!

**Reply by txoof** [reply]
@thoughtfix I've been working on a slideshow plugin over at the newer version of this project. You can find the newer version here:
https://github.com/txoof/paperpi/

I'll push a sample in the next day or so once I've cleaned up the code and fixed a few problems.

Here's a sample:
![image](https://user-images.githubusercontent.com/14262505/202815754-3f914586-f51d-42a5-9eaf-6cd2faa974aa.png)

**Comment by txoof** [comment]
@thoughtfix I've got a working version of PaperPi with a slideshow module ready to test out if you're interested.

It's on the `slideshow` branch. You can clone it and try it out at this link: https://github.com/txoof/PaperPi/tree/slideshow

Clone the repo and then run `PaperPi/utilities/create_devel_environment.sh`

The script will prompt you to install some libraries/dependencies. Once that's done, you can run PaperPi directly by running: `pipenv run paperpi/paperpi.py`

The slideshow simply pulls images from a local directory in either random or sequential order. It can handle adding and deleting images in the directory and will try to gracefully move to the next image if possible. 

Any issues or feedback are greatly appreciated!

**Comment by thoughtfix** [comment]
Good afternoon! I was missing some apt dependencies but finally got around to getting this up and running.  I had to add the data and slideshow photo list to ~/.config/com.txoof.paperpi/paperpi.ini  and got a pair of `dictionary pickle has bad data` errors. 

On reboot, it rolled back to the original and simply adding the slideshow didn't work for me. I decided to look at https://raw.githubusercontent.com/txoof/PaperPi/main/install/remote_install.sh and edit PP_GITBRANCH="main" to "slideshow" and ran it. Works great, and the recipient of the new e-ink photo frame is very happy.

Thank you!

**Reply by txoof** [reply]
The `dictionary pickle` errors are a bit misleading. It just means that the it couldn't load data that records the last image shown.

I should probably get rid of those.

I've done some considerable work on the slideshow plugin and it now includes optional "frames" around the images. You can see some samples on the [development branch](https://github.com/txoof/PaperPi/blob/development/paperpi/plugins/slideshow/README.md#additional-plugin-information)

After I test it a bit more, I'll merge it into the main branch. That would be a much more reliable place to install from :)

**Reply by txoof** [reply]
I cleaned up those pickle errors. 

If you switch over to the development branch, I'd appreciate your feedback on anything that's not working or that's weird.

**Comment by txoof** [comment]
consider merging some of the ideas from this in as a DeviantArt plugin: https://github.com/Martenfur/magic_frame
